File: examples/part3-loganalyzer-py/loganalyzer.py
"""Log analyzer — Python implementation of the Part 3 project. The same tool is
built in Rust, Go, and Python and benchmarked head-to-head.

Run the tests / the tool:

    python3 -m unittest discover -s . -p 'test_*.py'
    python3 loganalyzer.py access.log
"""

from __future__ import annotations

import sys
from collections import Counter
from dataclasses import dataclass, field


@dataclass(frozen=True)
class Entry:
    """One successfully parsed request line."""

    path: str
    status: int


def parse_line(line: str) -> Entry | None:
    """Parses one log line; None means malformed (counted, never fatal)."""
    parts = line.split('"', 2)
    if len(parts) != 3:
        return None
    req_fields = parts[1].split()
    suf_fields = parts[2].split()
    if len(req_fields) < 2 or not suf_fields:
        return None
    try:
        status = int(suf_fields[0])
    except ValueError:
        return None
    if not 100 <= status <= 599:
        return None
    return Entry(path=req_fields[1], status=status)


@dataclass
class Stats:
    """Aggregated statistics over a whole log."""

    total: int = 0
    malformed: int = 0
    by_class: list[int] = field(default_factory=lambda: [0] * 5)
    paths: Counter[str] = field(default_factory=Counter)

    def add_line(self, line: str) -> None:
        entry = parse_line(line)
        if entry is None:
            self.malformed += 1
            return
        self.total += 1
        self.by_class[entry.status // 100 - 1] += 1
        self.paths[entry.path] += 1

    def error_rate(self) -> float:
        if self.total == 0:
            return 0.0
        return (self.by_class[3] + self.by_class[4]) / self.total * 100

    def top_paths(self, n: int) -> list[tuple[str, int]]:
        # count desc, then path asc — deterministic across implementations.
        return sorted(self.paths.items(), key=lambda kv: (-kv[1], kv[0]))[:n]


def report(text: str) -> str:
    """Analyzes a whole log text and renders the shared output format."""
    stats = Stats()
    for line in text.splitlines():
        if line.strip():
            stats.add_line(line)
    out = [f"total: {stats.total}"]
    for i, label in enumerate(["1xx", "2xx", "3xx", "4xx", "5xx"]):
        out.append(f"{label}: {stats.by_class[i]}")
    out.append(f"malformed: {stats.malformed}")
    out.append(f"error_rate: {stats.error_rate():.1f}%")
    out.append("top paths:")
    for path, count in stats.top_paths(3):
        out.append(f"  {path}: {count}")
    return "\n".join(out) + "\n"


def main() -> None:
    if len(sys.argv) > 1:
        with open(sys.argv[1], encoding="utf-8") as f:
            text = f.read()
    else:
        text = sys.stdin.read()
    print(report(text), end="")


if __name__ == "__main__":
    main()

File: examples/part3-loganalyzer-go/loganalyzer.go
// Package main is the Go implementation of the Part 3 log-analyzer project.
// The same tool is built in Rust, Go, and Python and benchmarked head-to-head.
//
//	go test ./...
//	go build && ./loganalyzer access.log
package main

import (
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
	"strings"
)

// Entry is one successfully parsed request line.
type Entry struct {
	Path   string
	Status int
}

// ParseLine parses one log line; ok=false means malformed (counted, not fatal).
func ParseLine(line string) (Entry, bool) {
	parts := strings.SplitN(line, `"`, 3)
	if len(parts) != 3 {
		return Entry{}, false
	}
	reqFields := strings.Fields(parts[1])
	if len(reqFields) < 2 {
		return Entry{}, false
	}
	sufFields := strings.Fields(parts[2])
	if len(sufFields) < 1 {
		return Entry{}, false
	}
	status, err := strconv.Atoi(sufFields[0])
	if err != nil || status < 100 || status > 599 {
		return Entry{}, false
	}
	return Entry{Path: reqFields[1], Status: status}, true
}

// Stats aggregates a whole log.
type Stats struct {
	Total     uint64
	Malformed uint64
	ByClass   [5]uint64 // index 0 => 1xx ... 4 => 5xx
	Paths     map[string]uint64
}

// NewStats builds an empty aggregate.
func NewStats() *Stats { return &Stats{Paths: make(map[string]uint64)} }

// AddLine folds one line into the stats.
func (s *Stats) AddLine(line string) {
	entry, ok := ParseLine(line)
	if !ok {
		s.Malformed++
		return
	}
	s.Total++
	s.ByClass[entry.Status/100-1]++
	s.Paths[entry.Path]++
}

// ErrorRate is the percentage of valid requests that were 4xx/5xx.
func (s *Stats) ErrorRate() float64 {
	if s.Total == 0 {
		return 0
	}
	return float64(s.ByClass[3]+s.ByClass[4]) / float64(s.Total) * 100
}

type pathCount struct {
	Path  string
	Count uint64
}

// TopPaths returns the n most-requested paths, count desc then path asc.
func (s *Stats) TopPaths(n int) []pathCount {
	pairs := make([]pathCount, 0, len(s.Paths))
	for p, c := range s.Paths {
		pairs = append(pairs, pathCount{p, c})
	}
	sort.Slice(pairs, func(i, j int) bool {
		if pairs[i].Count != pairs[j].Count {
			return pairs[i].Count > pairs[j].Count
		}
		return pairs[i].Path < pairs[j].Path
	})
	if n > len(pairs) {
		n = len(pairs)
	}
	return pairs[:n]
}

// Report analyzes a whole log text and renders the shared output format.
func Report(input string) string {
	stats := NewStats()
	for _, line := range strings.Split(input, "\n") {
		if strings.TrimSpace(line) != "" {
			stats.AddLine(line)
		}
	}
	var b strings.Builder
	fmt.Fprintf(&b, "total: %d\n", stats.Total)
	for i, label := range []string{"1xx", "2xx", "3xx", "4xx", "5xx"} {
		fmt.Fprintf(&b, "%s: %d\n", label, stats.ByClass[i])
	}
	fmt.Fprintf(&b, "malformed: %d\n", stats.Malformed)
	fmt.Fprintf(&b, "error_rate: %.1f%%\n", stats.ErrorRate())
	b.WriteString("top paths:\n")
	for _, pc := range stats.TopPaths(3) {
		fmt.Fprintf(&b, "  %s: %d\n", pc.Path, pc.Count)
	}
	return b.String()
}

func main() {
	var data []byte
	var err error
	if len(os.Args) > 1 {
		data, err = os.ReadFile(os.Args[1])
	} else {
		data, err = io.ReadAll(os.Stdin)
	}
	if err != nil {
		fmt.Fprintln(os.Stderr, "loganalyzer:", err)
		os.Exit(1)
	}
	fmt.Print(Report(string(data)))
}
